pkg/core/image_mapper.go

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2018 The original author or authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package core
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+)
+
+const (
+	dockerHubHost     = "docker.io"
+	fullDockerHubHost = "index.docker.io"
+)
+
+// imageMapper does substring replacement of images with minimal delimiter checking.
+// It does not understand the full (YAML) syntax of the input string.
+// It certainly does not understand the semantics of the input string.
+type imageMapper struct {
+	mappings map[string]string
+}
+
+func newImageMapper(mappedHost string, mappedUser string, images []string) (*imageMapper, error) {
+	if err := contains(mappedHost, "/", `"`, " "); err != nil {
+		return nil, fmt.Errorf("Invalid registry hostname: %v", err)
+	}
+	if err := contains(mappedUser, "/", `"`, " "); err != nil {
+		return nil, fmt.Errorf("Invalid user: %v", err)
```

Comment:
Same as above, maybe there is an even better contract to enforce

Reply:
There's [that](https://docs.docker.com/engine/reference/commandline/tag/), as a starting point.

Reply:
Hmmm. The relevant text seems to be:
> Name components may contain lowercase letters, digits and separators. A separator is defined as a period, one or two underscores, or one or more dashes. A name component may not start or end with a separator.

But I'm not sure I trust this as then user names could contain separators, whereas
https://success.docker.com/article/how-do-you-register-for-a-docker-id states:
> Your [username that will become your Docker ID] must be between 4 and 30 characters long, and can only contain numbers and lowercase letters.